vllm/core/scheduler.py

```diff
@@ -841,6 +841,31 @@ def _schedule_priority_preemption(
         self.running = running_queue
         return force_preemption_count
 
+    def _configure_multi_step_accounting_for_best_of(self) -> None:
+        """Configure per-step multi-step scheduling, accounting for best_of>1
+
+        This method must be called each step in order to update scheduler config
+        `is_multi_step` member.
+
+        Default behavior: `is_multi_step = user_is_multi_step`
+        (i.e. match user-requested behavior.)
+        
+        But `is_multi_step = False` if any state queues contain
+        :class:`SequenceGroup`'s with `best_of>1` which is not
+        supported by multi-step.
+        """
+        if self.scheduler_config.user_is_multi_step:
+            self.scheduler_config.is_multi_step = all([
+                (sg.sampling_params is None
+                 or sg.sampling_params.best_of is None
+                 or sg.sampling_params.best_of < 2)
+                for state_queue in [self.waiting, self.running, self.swapped]
+                for sg in state_queue
+            ])
```

Comment:
I'm a bit concern about the overhead of this logic. It's on the latency critical path but looks quite expensive. Ideally we should check this after scheduling a batch so that we only need to check the sequences in the batch, but it is not compatible with the current design as we allocate block tables (including lookahead lots) when scheduling sequences.

Reply:
I agree. I think what I might be able to do is

1. Have the scheduler maintain a count of the number of "in progress" requests which have sampling params that are incompatible with multi-step
2. Scheduler decrements the counter when a SequenceGroup with incompatible sampling params is finished; the counter is always non-negative.
3. If the counter is non-zero that signals multi-step override

Reply:
Update: implemented lower-overhead logic.

* `Scheduler` maintains a counter of the number of unfinished `SequenceGroup` 's with `best_of>1` (i.e. seq groups incompatible with multi-step)
* `Scheduler.add_seq_group()` increments this counter if a new request has introduced a `seq_group` with `best_of>1`
* `Scheduler.free_finished_seq_groups()` decrements this counter for each finished `seq_group` which had `best_of>1`
* In an engine with `engine_permits_multi_step_scheduling=True`, `current_step_is_multi_step=True` in a given step *only* if the counter value is 0 (no incompatible seq groups)

Reply:
This logic is cheaper because the number of checks is (loosely speaking) O(# new seq groups + # finished seq groups), whereas the previous impl was O(all unfinished seq groups)